lib/azure/armrest/model/storage_account.rb

```diff
@@ -194,27 +193,43 @@ def directory_metadata(share, directory, key = access_key, options = {})
       # * marker
       # * maxresults
       # * timeout
+      # * all
+      #
+      # If the :all option is set to true, then this method will collect all
+      # records. Otherwise, it is capped at 5000 records by Azure, or whatever
+      # you set :maxresults to. If you set both the :all option and the :maxresults
+      # option, then all records will be collected in :maxresults batches.
       #
       def files(share, key = access_key, options = {})
         raise ArgumentError, "No access key specified" unless key
 
-        query = {:restype => 'directory', :comp => 'list'}.merge(options).to_query
+        query = "comp=list&restype=directory"
+
+        skip_defs = options[:skip_accessors_definition]
+
+        options.each do |okey, ovalue|
+          unless okey == :skip_accessors_definition
+            query += "&#{okey}=#{[ovalue].flatten.join(',')}"
+          end
+        end
 
         response = file_response(key, query, 'get', nil, share)
 
-        doc = Nokogiri::XML(response.body)
+        hash = Hash.from_xml(response.body)['EnumerationResults']['Entries']
         results = []
 
-        doc.xpath('//EnumerationResults/Entries').each do |element|
-          element.xpath('//Directory').each do |dir|
-            results << ShareDirectory.new(Hash.from_xml(dir.to_s)['Directory'])
-          end
-          element.xpath('//File').each do |file|
-            results << ShareFile.new(Hash.from_xml(file.to_s)['File'])
-          end
+        if hash && hash['Directory']
+          Array.wrap(hash['Directory']).each { |dir| results << ShareDirectory.new(dir, skip_defs) }
```

Comment:
Does `hash` always contains `Directory` and `File`?

Reply:
No, but the Array.wrap will convert nil to `[]` if it's non-existent.